plexpy/webserve.py

```diff
@@ -1915,21 +1933,98 @@ def get_item_children(self, rating_key='', **kwargs):
 
     @cherrypy.expose
     @requireAuth()
-    def pms_image_proxy(self, img='', width='0', height='0', fallback=None, **kwargs):
+    def pms_image_proxy(self, img='', ratingkey=None, width='0', height='0', fallback=None, **kwargs):
+        """ Grabs the images from pms and saved them to disk """
 
-        pms_connect = pmsconnect.PmsConnect()
-        result = pms_connect.get_image(img, width, height, fallback)
+        if not img or ratingkey:
+            logger.debug('No image input received')
+            return
 
-        if result:
-            cherrypy.response.headers['Content-type'] = result[1]
-            return result[0]
-        else:
-            return None
+        if ratingkey and not img:
+            img = '/library/metadata/%s/thumb/1337' % ratingkey
+
+        i = img.rsplit('/', 1)
+        i += [width, height]
+        img_string = '/'.join(i)
+        fp = hashlib.md5(img_string).hexdigest()
+        fp += '.jpg'  # we want to be able to preview the thumbs
+        c_dir = os.path.join(plexpy.CONFIG.CACHE_DIR, 'images')
+        ffp = os.path.join(c_dir, fp)
+
+        if not os.path.exists(c_dir):
+            os.mkdir(c_dir)
+
+        try:
+            if 'indexes' in img:
+                raise
+            return serve_file(path=ffp, content_type='image/jpeg')
+
+        except NotFound:
+            # the image does not exist, download it from pms
+            try:
+                pms_connect = pmsconnect.PmsConnect()
+                result = pms_connect.get_image(img, width, height)
+                cherrypy.response.headers['Content-type'] = result[1]
+
+                if result and result[0]:
+                    if 'indexes' not in img:
+                        with open(ffp, 'wb') as f:
+                            f.write(result[0])
```

Comment:
Isn't it still possible that this will cache a smaller image? e.g. the small thumbnail from the popup on the history table, then serve that smaller image for the larger posters elsewhere?